polybench/2mm.cpp

```diff
@@ -1,176 +1,159 @@
 			C[i * NL + j] = ((DATA_TYPE)i * (j + 3)) / NL;
 		}
 	}
 
-	for(i = 0; i < NI; i++) {
-		for(j = 0; j < NL; j++) {
+	for(int i = 0; i < NI; i++) {
+		for(int j = 0; j < NL; j++) {
 			D[i * NL + j] = ((DATA_TYPE)i * (j + 2)) / NK;
 		}
 	}
 }
 
-void mm2_cpu(DATA_TYPE* A, DATA_TYPE* B, DATA_TYPE* C, DATA_TYPE* D, DATA_TYPE* E) {
-	int i, j, k;
+void mm2_cpu(DATA_TYPE* A, DATA_TYPE* B, DATA_TYPE* C, DATA_TYPE* D, DATA_TYPE* E, size_t size) {
+	const int NI = (int)size;
+	const int NJ = (int)size;
+	const int NK = (int)size;
+	const int NL = (int)size;
 
-	for(i = 0; i < NI; i++) {
-		for(j = 0; j < NJ; j++) {
-			for(k = 0; k < NK; ++k) {
+	for(int i = 0; i < NI; i++) {
+		for(int j = 0; j < NJ; j++) {
+			for(int k = 0; k < NK; ++k) {
 				C[i * NJ + j] += A[i * NK + k] * B[k * NJ + j];
 			}
 		}
 	}
 
-	for(i = 0; i < NI; i++) {
-		for(j = 0; j < NL; j++) {
-			for(k = 0; k < NJ; ++k) {
+	for(int i = 0; i < NI; i++) {
+		for(int j = 0; j < NL; j++) {
+			E[i * NL + j] = 0;
+			for(int k = 0; k < NJ; ++k) {
 				E[i * NL + j] += C[i * NJ + k] * D[k * NL + j];
 			}
 		}
 	}
 }
 
-int main(int argc, char* argv[]) {
-	if(argc >= 2) {
-		const auto problem_size = std::atoi(argv[1]);
-		NI = problem_size;
-		NJ = problem_size;
-		NK = problem_size;
-		NL = problem_size;
-	}
-	std::cout << "Problem size: " << NI << "\n";
-
-	std::vector<DATA_TYPE> A(NI * NK);
-	std::vector<DATA_TYPE> B(NK * NJ);
-	std::vector<DATA_TYPE> C(NI * NJ);
-	std::vector<DATA_TYPE> D(NJ * NL);
-	std::vector<DATA_TYPE> E(NI * NL);
+class Polybench_2mm {
+  public:
+	Polybench_2mm(const BenchmarkArgs& args) : args(args), size(args.problem_size) {}
 
-	init_array(A.data(), B.data(), C.data(), D.data());
+	void setup() {
+		A.resize(size * size);
+		B.resize(size * size);
+		C.resize(size * size);
+		D.resize(size * size);
+		E.resize(size * size);
 
-	if(shouldDoCpu()) {
-		double t_start = rtclock();
-		mm2_cpu(A.data(), B.data(), C.data(), D.data(), E.data());
-		double t_end = rtclock();
-		fprintf(stdout, "CPU Runtime: %0.6lfs\n", t_end - t_start);
+		init_array(A.data(), B.data(), C.data(), D.data(), size);
 	}
 
-	{
+	void run() {
 		using namespace cl::sycl;
 
-		cl::sycl::queue queue;
-
-		buffer<DATA_TYPE, 2> A_buffer(range<2>(NK, NI));
-		initDeviceBuffer(queue, A_buffer, A.data());
-
-		buffer<DATA_TYPE, 2> B_buffer(range<2>(NJ, NK));
-		initDeviceBuffer(queue, B_buffer, B.data());
-
-		buffer<DATA_TYPE, 2> C_buffer(range<2>(NJ, NI));
-		initDeviceBuffer(queue, C_buffer, C.data());
-
-		buffer<DATA_TYPE, 2> D_buffer(range<2>(NL, NJ));
-		initDeviceBuffer(queue, D_buffer, D.data());
-
-		buffer<DATA_TYPE, 2> E_buffer(range<2>(NL, NI));
+		buffer<DATA_TYPE, 2> A_buffer(A.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> B_buffer(B.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> C_buffer(C.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> D_buffer(D.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> E_buffer(E.data(), range<2>(size, size));

```

Comment:
Be advised, the framework will perform a loop over `{setup(); start_timer(); run(); stop_timer()}`, averaging the runtimes. Creating buffers here will cause data transfers to be included in the timings, which may or may not be what you want.
If only the kernel should be measured, you can create the buffers in `setup()` and force copying them to device with a dummy kernel. Or create them as member variables of the class and hope that the initial data transfer in the first run will be averaged away.